rts/game_MC/lua/unit_generator.lua

```diff
@@ -59,3 +60,187 @@ function rts_unit_generator.generate(proxy, num_players, seed)
   proxy:send_cmd_change_player_resource(0, 1500)
   proxy:send_cmd_change_player_resource(1, 100)
 end
+
+function can_pass(proxy, x, y)
+  local X = proxy:get_x_size()
+  local Y = proxy:get_y_size()
+  if not (x >= 0 and x < X and y >= 0 and y < Y) then
+    return false
+  end
+  local ty = proxy:get_slot_type(x, y, 0)
+  return not (ty == Terrain.WATER or ty == Terrain.ROCK)
+end
+
+
+function can_reach(proxy, sx, sy, tx, ty)
+  local X = proxy:get_x_size()
+  local Y = proxy:get_y_size()
+  local dx = {-1, -1, 1, 1}
+  local dy = {-1, 1, -1, 1}
+
+  function dfs(x, y, tx, ty, visited)
+    if x == tx and y == ty then
+      return true
+    end
+    visited[x * Y + y] = true
+    for i = 1, #dx do
+      local nx = x + dx[i]
+      local ny = y + dy[i]
+      if nx >= 0 and nx < X and ny >= 0 and ny < Y and visited[nx * Y + ny] == nil then
+        if can_pass(proxy, nx, ny) and dfs(nx, ny, tx, ty, visited) then
+          return true
+        end
+      end
+    end
+    return false
+  end
+  return dfs(sx, sy, tx, ty, {})
+end
+
+
+function add_base_and_workers(proxy, x, y, player_id)
+  local X = proxy:get_x_size()
+  local Y = proxy:get_y_size()
+  local n_workers = 3
+  proxy:send_cmd_create(UnitType.TOWN_HALL, __make_p(x, y), player_id, 0)
+  local taken = {}
+  taken[x * Y + y] = true
+  for i = 1, n_workers do
+    local found = false
+    while not found do
+      local xx = x + math.random(-5, 5)
+      local yy = y + math.random(-5, 5)
+      if can_pass(proxy, xx, yy) and taken[xx * Y + yy] == nil then
+        proxy:send_cmd_create(UnitType.PEASANT, __make_p(xx, yy), player_id, 0)
+        taken[xx * Y + yy] = true
+        found = true
+      end
+    end
+  end
+  local n_resource = 1
```

Comment:
one thing to consider is that we may need to control the distance from town hall to resource. e.g. 2 distance away will be roughly 2X efficiency than 4, and is often the deciding factor from experience in old minirts